### PR TITLE
[Fix #3058] Add `Style/SpaceInsideArrayPercentLiteral` and `Style/SpaceInsidePercentLiteralDelimiters` cops

### LINKED ISSUE #3058: Doesn't remove space from %w array
for an array defined like 
`%w( 2016-04-01 2016-03-01 2016-02-01 2016-01-01)`

Rubocop won't fix the initial space (or add a space at the end)

Using latest rubocop

### PR DESCRIPTION
**Pull request by owst**
Add `Style/SpaceInsideArrayPercentLiteral` cop:
```
      # Checks for unnecessary additional spaces inside array percent literals
      # (i.e. %i/%w).
      #
      # @good
      # %i(foo bar baz)
      #
      # @bad
      # %w(foo  bar  baz)
```
Add `Style/SpaceInsidePercentLiteralDelimiters` cop:
```
      # Checks for unnecessary additional spaces inside the delimiters of
      # %i/%w/%x literals.
      #
      # @good
      # %i(foo bar baz)
      #
      # @bad
      # %w( foo bar baz )
      #
      # @bad
      # %x(  ls -l )
```
-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it)
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -755,6 +755,10 @@ Style/SpaceAroundOperators:
   StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#spaces-operators'
   Enabled: true
 
+Style/SpaceInsideArrayPercentLiteral:
+  Description: 'No unnecessary additional spaces inside %i/%w literals. '
```
Comment (bbatsov) [inline config/enabled.yml]: There's an extra space here. :-)

File: spec/rubocop/cop/style/space_inside_array_percent_literal_spec.rb [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                               "  %#{type}(",
+                               '    a',
+                               '    b',
+                               '    c',
+                               '  )',
+                               'end'])
+          expect(cop.messages).to be_empty
+        end
+
+        it 'accepts newlines and additional following alignement spaces' do
```
Comment (bbatsov) [inline spec/rubocop/cop/style/space_inside_array_percent_literal_spec.rb]: There's a typo here.

**Comment by bbatsov** [comment]
This cop is effectively checking for 2 different things, which is typically not the granularity we're aiming for. I think the first check makes sense for pretty much every %-literal. (except maybe `%` and `%s`, although I'd definitely don't see much value in padding strings/symbols with spaces there).

**Comment by owst** [comment]
I've fixed the typo and additional space @bbatsov, good spot! 

I agree about splitting the cop in two, but I'm not sure we want to check for unnecessary leading/trailing space inside other %-literals (except maybe `%x(...)`) since it is significant (part of the literals):

| %-literal | Additional leading/trailing whitespace significant |
| ----------------- | -------------------------------------------- |
| `%(...)` - String | yes |
| `%q(...)` - String | yes |
| `%i(...)` - Array of Symbols | no  (unless escaped with `\`) |
| `%r(...)` - Regexp | yes |
| `%s(...)` - Symbol | yes |
| `%w(...)` - Array of Strings | no  (unless escaped with `\`) |
| `%x(...)` - Backtick | no |

For a contrived example, consider: 
```
> ruby -e 'p %r{ a } =~ "a"'
nil
> ruby -e 'p %r{a} =~ "a"'
0
```

So I propose to split the cop into two cops: 
1. `SpaceInsideArrayPercentLiteral` for detecting > 1 space between entries in `%i/%w`
2. `SpaceInsidePercentLiteralDelimiters` for detecting unnecessary leading/trailing space inside `%i/%w/%x)`

What do you think?

**Comment by bbatsov** [comment]
> I agree about splitting the cop in two, but I'm not sure we want to check for unnecessary leading/trailing space inside other %-literals (except maybe %x(...)) since it is significant (part of the literals):

I'm aware of this, but I find using %-something for a thing padded with spaces to be a bit hard to read/mentally parse and I've done this myself.

> What do you think?

Fine by me.

**Comment by owst** [comment]
@bbatsov changes made, please take a look when you are able to!

**Comment by bbatsov** [comment]
Looks OK, but has to be rebased. Maybe we add something about this in the style guide as well.

**Comment by owst** [comment]
Rebased @bbatsov - I've made a note to myself to submit a change to the style guide about this too.

**Pull request merged**